File: Scala/src/Sources/Count/count_pn.scala
package Sources.Count

import Others.value._
import scala.collection.mutable.ArrayBuffer

object count_pn {
    private def count_sort_pn(a: Array[Int]): Unit = {
        var s, l = a(0)
        for (i <- a.indices) {
            if (a(i) <= s)
                s = a(i)
            if (a(i) >= l)
                l = a(i)
        }

        val b = ArrayBuffer.fill((l - s) + 1)(0)

        for (i <- a.indices)
            b(a(i) - s) += 1

        var k = 0
        for (i <- s until l + 1) {
            for (_ <- 0 until b(i - s)) {
                a(k) = i
                k += 1
            }
        }
    }

    def main(args: Array[String]): Unit = {
        val s = 1000000 // Changeable size
        val a = rand_pn(s)

        val start = System.currentTimeMillis()

        count_sort_pn(a)

        val end = System.currentTimeMillis() - start

        for (i <- a.indices)
            println(a(i))
        print(end + " ms")
    }
}

File: Kotlin/src/main/kotlin/Sources/Count/count_pn.kt
package Sources.Count


fun count_pn(a: IntArray) {
    var s = a[0]
    var l = a[0]
    for (i in a.indices) {
        if (a[i] <= s)
            s = a[i]
        if (a[i] >= l)
            l = a[i]
    }

    val b: MutableList<Int> = MutableList((l-s)+1){0}

    for (i in a.indices)
        ++b[a[i]-s]

    var k=0
    for (i in s until l+1) {
        for (j in 0 until b[i-s])
            a[k++] = i
    }
}

fun main() {
    val a = intArrayOf(2, 1, 0, -1, -2)

    val start = System.currentTimeMillis()

    count_pn(a)

    val end = System.currentTimeMillis() - start

    for (i in a.indices)
        println(a[i])
    print("$end ms")
}